src/style-spec/expression/definitions/number_format.js

```diff
@@ -0,0 +1,114 @@
+// @flow
+
+import { StringType, NumberType } from '../types';
+
+import type { Expression } from '../expression';
+import type EvaluationContext from '../evaluation_context';
+import type ParsingContext from '../parsing_context';
+import type { Type } from '../types';
+
+declare var Intl: {
+    NumberFormat: Class<Intl$NumberFormat>
+};
+
+declare class Intl$NumberFormat {
+    constructor (
+        locales?: string | string[],
+        options?: NumberFormatOptions
+    ): Intl$NumberFormat;
+
+    static (
+        locales?: string | string[],
+        options?: NumberFormatOptions
+    ): Intl$NumberFormat;
+
+    format(a: number): string;
+
+    resolvedOptions(): any;
+}
+
+type NumberFormatOptions = {
+    style?: 'decimal' | 'currency' | 'percent';
```

Comment:
`NSNumberFormatterStyle` actually supports a number of styles, including `NSNumberFormatterSpellOutStyle` and `NSNumberFormatterOrdinalStyle`. However, I recognize the need to stick to JavaScript’s `Intl` as a common denominator, without implementing our own custom types that could conflict with future JavaScript types.

This does mean that the style author will need to hard-code some logic to implement [short formats](https://www.unicode.org/cldr/charts/26/by_type/numbers.compact_decimal_formatting.html#Short_Formats_), such as the “1k” for 1,000 that Supercluster implements (only for English) in the `point_count_abbreviated` property (mapbox/supercluster#110).

/ref tc39/ecma402#37 tc39/ecma402#215.